lib/src/main/java/com/squareup/cycler/Update.kt

```diff
@@ -98,16 +136,24 @@ class Update<I : Any>(private val oldRecyclerData: RecyclerData<I>) {
   }
 
   /**
-   * Result object for [generateUpdateWork]. It contains:
+   * Result object for [generateUpdateLambdas]. It contains:
+   * @param populateMode [UpdateMode] for updating the list with its initial dataset
+   * @param replaceMode [UpdateMode] for replacing an existing dataset
+   * @param populating Whether this update is updating the list with its initial dataset
+   * @param replacing Whether this update is replacing an existing dataset
    * @param asyncWork List of lambdas to be executed asynchronously (can be empty).
    * @param notifications List of lambdas receiving the adapter that will notify of the changes.
    */
   internal class UpdateWork(
-    private val allowSynchronousUpdate: Boolean,
+    val populateMode: UpdateMode,
+    val replaceMode: UpdateMode,
+    val populating: Boolean,
+    val replacing: Boolean,
     val asyncWork: List<() -> Unit>,
     val notifications: List<(Adapter<*>) -> Unit>
   ) {
-    fun isSynchronous(): Boolean = allowSynchronousUpdate && asyncWork.isEmpty()
+    fun isSynchronous(): Boolean = (populating && populateMode is Sync) ||
+        (replacing && replaceMode is Sync)
```

Comment:
This is used to tell Recycler not to launch a coroutine.
Maybe we could instead have two properties instead of mode:
type: populate / replace / add
mode: sync / async

That'll be easier to read here, and we do whatever this says.